<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Mathew Chamberlain" />

<meta name="date" content="2021-02-26" />

<title>Signac and SPRING: Learning CD56 NK cells from multi-modal analysis of CITE-seq PBMCs from 10X Genomics</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Signac and SPRING: Learning CD56 NK cells from multi-modal analysis of CITE-seq PBMCs from 10X Genomics</h1>
<h4 class="author">Mathew Chamberlain</h4>
<h4 class="date">2021-02-26</h4>



<p>This vignette shows how to use SignacX with Seurat and SPRING to learn a new cell type category from single cell data.</p>
<div id="load-data" class="section level2">
<h2>Load data</h2>
<p>We start with CITE-seq data that were already classified with SignacX using the SPRING pipeline.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">library</span>(SignacX)</span></code></pre></div>
<p>Load CITE-seq data from 10X Genomics processed with SPRING and classified with SignacX already.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># load CITE-seq data</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>data.dir =<span class="st"> &#39;./CITESEQ_EXPLORATORY_CITESEQ_5K_PBMCS/FullDataset_v1_protein&#39;</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>E =<span class="st"> </span><span class="kw">CID.LoadData</span>(<span class="dt">data.dir =</span> data.dir)</span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="co"># Load labels</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>json_data =<span class="st"> </span>rjson<span class="op">::</span><span class="kw">fromJSON</span>(<span class="dt">file=</span><span class="kw">paste0</span>(data.dir,<span class="st">&#39;/categorical_coloring_data.json&#39;</span>))</span></code></pre></div>
<p>Create a Seurat object for the protein expression data; we will use this as a reference.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># separate protein and gene expression data</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>logik =<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;Total&quot;</span>, <span class="kw">rownames</span>(E))</span>
<span id="cb3-3"><a href="#cb3-3"></a>P =<span class="st"> </span>E[logik,]</span>
<span id="cb3-4"><a href="#cb3-4"></a>E =<span class="st"> </span>E[<span class="op">!</span>logik,]</span>
<span id="cb3-5"><a href="#cb3-5"></a></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co"># CLR normalization in Seurat</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="kw">colnames</span>(P) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(P)</span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="kw">colnames</span>(E) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(E)</span>
<span id="cb3-9"><a href="#cb3-9"></a>reference &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(E)</span>
<span id="cb3-10"><a href="#cb3-10"></a>reference[[<span class="st">&quot;ADT&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">CreateAssayObject</span>(<span class="dt">counts =</span> P)</span>
<span id="cb3-11"><a href="#cb3-11"></a>reference &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(reference, <span class="dt">assay =</span> <span class="st">&quot;ADT&quot;</span>, <span class="dt">normalization.method =</span> <span class="st">&quot;CLR&quot;</span>)</span></code></pre></div>
<p>Identify CD56 bright NK cells based on protein expression data.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># generate labels </span></span>
<span id="cb4-2"><a href="#cb4-2"></a>lbls =<span class="st"> </span>json_data<span class="op">$</span>CellStates<span class="op">$</span>label_list</span>
<span id="cb4-3"><a href="#cb4-3"></a>lbls[lbls <span class="op">!=</span><span class="st"> &quot;NK&quot;</span>] =<span class="st"> &quot;Unclassified&quot;</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>CD16 =<span class="st"> </span>reference<span class="op">@</span>assays<span class="op">$</span>ADT<span class="op">@</span>counts[<span class="kw">rownames</span>(reference<span class="op">@</span>assays<span class="op">$</span>ADT<span class="op">@</span>counts) <span class="op">==</span><span class="st"> &quot;CD16-TotalSeqB-CD16&quot;</span>,]</span>
<span id="cb4-5"><a href="#cb4-5"></a>CD56 =<span class="st"> </span>reference<span class="op">@</span>assays<span class="op">$</span>ADT<span class="op">@</span>counts[<span class="kw">rownames</span>(reference<span class="op">@</span>assays<span class="op">$</span>ADT<span class="op">@</span>counts) <span class="op">==</span><span class="st"> &quot;CD56-TotalSeqB-CD56&quot;</span>,]</span>
<span id="cb4-6"><a href="#cb4-6"></a>logik =<span class="st"> </span><span class="kw">log2</span>(CD56) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span> <span class="op">&amp;</span><span class="st"> </span><span class="kw">log2</span>(CD16) <span class="op">&lt;</span><span class="st"> </span><span class="fl">7.5</span> <span class="op">&amp;</span><span class="st"> </span>lbls <span class="op">==</span><span class="st"> &quot;NK&quot;</span>; <span class="kw">sum</span>(logik)</span>
<span id="cb4-7"><a href="#cb4-7"></a>lbls[logik] =<span class="st"> &quot;NK.CD56bright&quot;</span></span></code></pre></div>
</div>
<div id="signacx" class="section level2">
<h2>SignacX</h2>
<p>Generate a training data set from the reference data and save it for later use. Note:</p>
<ul>
<li>SignacBoot performs feature selection, bootstrapping, imputation and normalization to derive a training data set from single cell data.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># generate bootstrapped single cell data</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>R_learned =<span class="st"> </span><span class="kw">SignacBoot</span>(<span class="dt">E =</span> E, <span class="dt">spring.dir =</span> data.dir, <span class="dt">L =</span> <span class="kw">c</span>(<span class="st">&quot;NK&quot;</span>, <span class="st">&quot;NK.CD56bright&quot;</span>), <span class="dt">labels =</span> lbls, <span class="dt">logfc.threshold =</span> <span class="dv">1</span>)</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co"># save the training data</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="kw">save</span>(R_learned, <span class="dt">file =</span> <span class="st">&quot;training_NKBright_v207.rda&quot;</span>)</span></code></pre></div>
</div>
<div id="classify-a-new-data-set-with-the-model" class="section level2">
<h2>Classify a new data set with the model</h2>
<p>Load expression data for a different data set (this was also previously processed through SPRING and SignacX)</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># Classify another data set with new model</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="co"># load new data</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>new.data.dir =<span class="st"> &quot;./PBMCs_5k_10X/FullDataset_v1&quot;</span></span>
<span id="cb6-4"><a href="#cb6-4"></a>E =<span class="st"> </span><span class="kw">CID.LoadData</span>(<span class="dt">data.dir =</span> new.data.dir)</span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co"># load cell types identified with Signac</span></span>
<span id="cb6-6"><a href="#cb6-6"></a>json_data =<span class="st"> </span>rjson<span class="op">::</span><span class="kw">fromJSON</span>(<span class="dt">file=</span><span class="kw">paste0</span>(new.data.dir,<span class="st">&#39;/categorical_coloring_data.json&#39;</span>))</span></code></pre></div>
<p>Generate new labels. Note:</p>
<ul>
<li>Signac trains an ensemble of 100 neural network classifiers using the new training data set built above (R_learned), and then classifies unseen data (E).</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># generate new labels</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>cr_learned =<span class="st"> </span><span class="kw">Signac</span>(<span class="dt">E =</span> E, <span class="dt">R =</span> R_learned, <span class="dt">spring.dir =</span> new.data.dir)</span></code></pre></div>
<p>Now we amend the existing labels (classified previously with SignacX); we add the new labels and generate a new SPRING layout.Note:</p>
<ul>
<li>We usually copy the existing SPRING files from “FullDataset_v1” to “FullDataset_v1_Learned” to generate a new layout while preserving the existing layout.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># modify the existing labels</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>cr =<span class="st"> </span><span class="kw">lapply</span>(json_data, <span class="cf">function</span>(x) x<span class="op">$</span>label_list)</span>
<span id="cb8-3"><a href="#cb8-3"></a>logik =<span class="st"> </span>cr<span class="op">$</span>CellStates <span class="op">==</span><span class="st"> &#39;NK&#39;</span></span>
<span id="cb8-4"><a href="#cb8-4"></a>cr<span class="op">$</span>CellStates[logik] =<span class="st"> </span>cr_learned[logik]</span>
<span id="cb8-5"><a href="#cb8-5"></a>logik =<span class="st"> </span>cr<span class="op">$</span>CellStates_novel <span class="op">==</span><span class="st"> &#39;NK&#39;</span></span>
<span id="cb8-6"><a href="#cb8-6"></a>cr<span class="op">$</span>CellStates_novel[logik] =<span class="st"> </span>cr_learned[logik]</span>
<span id="cb8-7"><a href="#cb8-7"></a>new.data.dir =<span class="st"> </span><span class="kw">paste0</span>(new.data.dir, <span class="st">&quot;_Learned&quot;</span>)</span></code></pre></div>
<p>Save results</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># save</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>dat =<span class="st"> </span><span class="kw">CID.writeJSON</span>(cr, <span class="dt">spring.dir =</span> new.data.dir, <span class="dt">new_colors =</span> <span class="kw">c</span>(<span class="st">&#39;red&#39;</span>), <span class="dt">new_populations =</span> <span class="kw">c</span>( <span class="st">&#39;NK.CD56bright&#39;</span>))</span></code></pre></div>
<details>
<p><summary><strong>Session Info</strong></summary></p>
<pre><code>## R version 3.5.0 (2018-04-23)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS/LAPACK: /site/ne/home/i0369218/.local/share/r-miniconda/envs/r-reticulate/lib/libopenblasp-r0.3.10.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=C              
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## loaded via a namespace (and not attached):
##  [1] compiler_3.5.0  magrittr_1.5    formatR_1.7     htmltools_0.4.0
##  [5] tools_3.5.0     yaml_2.2.1      Rcpp_1.0.4.6    stringi_1.4.6  
##  [9] rmarkdown_2.1   highr_0.8       knitr_1.28      stringr_1.4.0  
## [13] digest_0.6.18   xfun_0.12       rlang_0.4.8     evaluate_0.14</code></pre>
</details>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
